### PR TITLE
Correct distinction between imported modules and all modules

### PR DESCRIPTION
**Pull request by ehildenb**
This PR fixes an issue that has shown up in https://github.com/runtimeverification/pyk/pull/217 due to duplicate productions for `<generatedCounter>` cell label. While we probably should not be duplicating that production, at least we can fix this by being more disciplined about how we treat modules in a K Definition for now. I've confirmed locally that this allows us to update to latest K.

- `KDefinition` class is modified in the following ways:
  - The field `modules` is renamed to `all_modules`
  - Cached property `module_names` is renamed to `all_module_names`.
  - Cached property `module_dict` is added for fast lookups of modules by names (and used in method `module(str) -> KFlatModule`).
  - Cached property `module_names` is added, which returns only the names of modules imported by the main module of the definition.
  - Cached property `modules` is added, which returns only the modules imported by the main module of the definition.
  - Most lookup methods/properties about the definition are modified to work over `KDefinition.modules` instead of `KDefinition.all_modules`.
  - A test is added which shows the difference between `KDefinition.modules` and `KDefinition.all_modules`.
- `KPrint` class is modified in the following ways:
  - When printing out definitions, we use `KDefinition.all_modules`. This has 2 effects, (i) that modules which are not imported into the main module are printed as well (desired for code generation purposes), and (ii) that definitions which use a `requires "..."` statement to import other modules, but don't actually have them in context, print out successfully.
  - When building the unparsing symbol-table, we use `KDefinition.all_modules`. This is because there are several symbols defined in `kast.md` that are needed for unparsing (like `#EmptyK`), which are not included in the definition usually.

I'd like to note that one change which was tricky for me to find was that in many places in `KDefinition`, we were relying on the default iterator for `KDefinition` to return `KDefinition.modules`. This ended up being hard for me to find the breakage, and I had to look through all occurances of `self` in the `outer.py` to determine if it needed to be changed to `self.modules` or `self.all_modules`. I'd like to be more explicit in the future, and avoid these types of ambiguities. It seems to me that having a default iterator is useful when you want to pass the object into some external algorithm that relies on that, but in our own codebase, we always should know the method/property/field name we want to call instead.

### CODE REVIEW

**Review by tothtamas28 — APPROVED** [review]
I suggest the following names:

```
modules, module_names, modules_by_name                             # <- what is now all_modules
imported_modules, imported_module_names, imported_modules_by_name  # <- what is now modules
```

This naming makes the difference between the two collections obvious.

The type of the dictionaries should be `FrozenDict`.

File: src/pyk/kast/outer.py [hunk 1]
```diff
@@ -951,46 +951,65 @@ def let(
...
     @cached_property
     def module_names(self) -> Tuple[str, ...]:
-        return tuple(module.name for module in self)
+        module_names = [self.main_module_name]
+        seen_modules = []
+        while len(module_names) > 0:
+            mname = module_names.pop(0)
+            if mname not in seen_modules:
+                seen_modules.append(mname)
+                module_names.extend([i.name for i in self.all_modules_dict[mname].imports])
```
[the remark this answers is not in the document]
Comment (tothtamas28) [reply src/pyk/kast/outer.py:979]: Also, are you sure `module_names` will only contain unique entries? Can you add a test for it?

[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kast/outer.py:979]: There is a test for it, it was failing until I added this on an integration test, because it was getting `SET` module imported twice. I'll add a direct unit test.

**Comment by ehildenb** [comment]
I want to use `all_modules` and `modules` instead of `modules` and `imported_modules` (which was my first solution), because I think by default most people will want just the imported modules, not all the modules.

File: src/pyk/kast/outer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kast/outer.py:979]: I've added a more detailed test that no duplicates occur in a diamond import.
